### PR TITLE
Openapi auth support with tests

### PR DESCRIPTION
**Pull request by praneetap**
*Issue #, if available:*

*Description of changes:*

*Description of how you validated changes:*

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by brettstack** [comment]
Looks good. Have you actually deployed a transformed template to CloudFormation and verified it sets up APIGW correctly?

**Comment by praneetap** [comment]
> Looks good. Have you actually deployed a transformed template to CloudFormation and verified it sets up APIGW correctly?

I was able to deploy a transformed version of this template successfully.
```
Globals:
  Api:
    OpenApiVersion: '3.0.1'

Resources:
  ImplicitApiFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: s3://praneep-test-2/thumbnails.zip
      Handler: index.gethtml
      Runtime: nodejs8.10
      Events:
        GetHtml:
          Type: Api
          Properties:
            Path: /
            Method: get
```

**Comment by brettstack** [comment]
Could you try with a template using all Authorizers similar to your test? We will write internal integ tests specifically for this but it would be good to verify before we get that far

File: tests/translator/test_translator.py [hunk 1]
```diff
@@ -321,6 +316,74 @@ def test_transform_success(self, testcase, partition_with_region):
...
+      [
+       ("aws", "ap-southeast-1"),
+       ("aws-cn", "cn-north-1"),
+       ("aws-us-gov", "us-gov-west-1")
+      ] # Run all the above tests against each of the list of partitions to test against
+      )
+    )
+    @patch('samtranslator.plugins.application.serverless_app_plugin.ServerlessAppPlugin._sar_service_call', mock_sar_service_call)
+    @patch('botocore.client.ClientEndpointBridge._check_default_region', mock_get_region)
+    def test_transform_success_openapi3(self, testcase, partition_with_region):
```
Comment (keetonian) [inline tests/translator/test_translator.py:339]: I like this idea. The other method was getting too long. I wonder if we can further improve this... but this looks good to me!

**Review by keetonian — APPROVED** [review]
Looks good!

**Review by keetonian — COMMENTED** [review]
I might like to see a test with: 
- OpenApi 3.0 specified in the swagger, but not in the OpenApiVersion property
- Paths defined in the swagger
- Authorizers

Just to ensure that the code adds authorizers correctly to existing swagger (for both swagger 2.0 and OpenApi 3.0) and we get that captured in the tests. Something similar to [this](https://github.com/awslabs/serverless-application-model/pull/932/files#diff-f6c9867e03ddb078f2caeb472cf5cef6) from your other PR.

**Comment by praneetap** [comment]
@keetonian will add these tests thanks! :)

**Comment by praneetap** [comment]
@brettstack sorry I thought we were on the other pr for some reason! yes I will try to deploy a template with auth here and post it soon. Thanks!

File: tests/translator/input/api_with_auth_all_minimum_openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+Globals:
```
Comment (praneetap) [inline tests/translator/input/api_with_auth_all_minimum_openapi.yaml:1]: @brettstack I could deploy this template (with valid s3 bucket). @keetonian This template should cover tests for #2 and #3 in your comment.

File: tests/translator/input/api_with_auth_no_openapi_flag.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
+Globals:
```
Comment (praneetap) [inline tests/translator/input/api_with_auth_no_openapi_flag.yaml]: @keetonian @brettstack I need your recommendation here - If I dont have a global api flag and define the inline swagger as openapi, we still add the auth parts to this inline swagger in swagger 2.0 format. This is obviously the wrong thing to do.
1. We can either fail if during auth we check that this is "openapi" but does not have OpenApiVersion:3
2. We see that this has "openapi: 3" and not "swagger: 2" so we add auth in openapi format (securitySchemes)

What is your recommendation?

**Comment by brettstack** [comment]
Currently we fail if we don't detect "swagger," however, we know that some developers add both "openapi" and "swagger" to get around this. Do we think we will potentially break customers who do this workaround? If so, we should fail with error message saying they need to specify OpenApiVersion 3 (also add code comment with this reason)

**Comment by praneetap** [comment]
> Currently we fail if we don't detect "swagger," however, we know that some developers add both "openapi" and "swagger" to get around this. Do we think we will potentially break customers who do this workaround? If so, we should fail with error message saying they need to specify OpenApiVersion 3 (also add code comment with this reason)

If that's a concern I can check that - 
if swagger, proceed with 'securityDefinitions'
else if openapi, check we have a OpenApiVersion flag. If not, throw error.


This way we won't break that customers with that workaround.

File: samtranslator/model/api/api_generator.py [hunk 4]
```diff
@@ -287,9 +287,28 @@ def _add_auth(self):
...
 
     def _openapi_auth_postprocess(self, definition_body):
+        """
+        Convert auth components to openapi 3 in definition body if OpenApiVersion flag is specified.
+
+        To maintain backward compatibility with existing swagger documents that contain both swagger
+        and openapi defined in the definition body, we return if there is "swagger" defined in the
+        definition body.
+
+        If the definition body contains "openapi" and not "swagger", we enforce that the OpenApiVersion
```
Comment (keetonian) [inline samtranslator/model/api/api_generator.py:301]: Why do we enforce this case?

Comment (praneetap) [reply samtranslator/model/api/api_generator.py:301]: > Why do we enforce this case?

Read Brett's comment above. Additionally, if we are generating/adding to the definitionBody, we make sure the OpenApiVersion flag is specified before we decide to go with OAS3.

**Pull request closed without merging**